### ISSUE TITLE
SocketError when running echoslam

### ISSUE THREAD

**Issue by andersastor** [tomasfarias/dota-gsi#2]
When I try to run echoslam with -gamestateintegration in my launch properties and the example configuration file supplied in the ReadMe I get two different kinds of SocketError depending on how I format the uri: 
`PS D:\Workspace\dota-gsi> .\target\debug\echoslam.exe --uri "http://127.0.0.1:3000"
Error: SocketError(Os { code: 11001, kind: Uncategorized, message: "Host not found." })`
`PS D:\Workspace\dota-gsi> .\target\debug\echoslam.exe --uri "http://127.0.0.1:3000/"
Error: SocketError(Error { kind: InvalidInput, message: "invalid port value" })`

Folders I have tried placing the cfg file:
* D:\SteamLibrary\steamapps\common\dota 2 beta\csgo\cfg (the one specified in the ReadMe for Windows)
* D:\SteamLibrary\steamapps\common\dota 2 beta\game\dota\cfg
* D:\SteamLibrary\steamapps\common\dota 2 beta\game\dota\cfg\gamestate_integration (folder used by DotaBuff and Overwolf, both of which work fine)

Any assistance would be appreciated.

**Comment by andersastor** [comment]
I've experimented a bit more, I tried using this .NET-library https://github.com/antonpup/Dota2GSI and it works fine with the cfg file in D:\SteamLibrary\steamapps\common\dota 2 beta\game\dota\cfg\gamestate_integration, but when I try to run echoslam with the same uri I get the above errors still. However I would really like to use this crate as I want to learn Rust.

**Comment by andersastor** [comment]
So I figured it out in the end, turns out you need to omit the http:// part of the uri when supplying it to the GSIServer object, and no matter if you specify localhost or 127.0.0.1 in your cfg file you must use 127.0.0.1 in the uri to the GSIServer. 

However, now that I have it running I have found multiple other problems. #1, many of the fields in GameState are private, such as Map and the game time which is what I wanted to access. #2, while the prints work, there are also a whole lot of error printouts clogging up the output window making me unsure whether the parsed objects are really up to date, as there are several seconds between each successful parse. The following is the error message that is presented: `[2023-05-23T15:17:55Z ERROR dota] Failed to parse JSON body: EOF while parsing a value at line 1 column 0`
EDIT: There are several seconds between each successful parse, the error messages comes from the parser getting empty JSONs several times per second which causees the error messages to be printed. Anything that can be done about this?

I am also missing a whole lot of documentation of how to actually use this, such as the two different run/run_with_handler which seem to do the exact same thing except run_with_handler clones the handler first.

Am I missing something or is this library just not that functional, and I should turn elsewhere? I would be willing to make a PR for the private fields, but if the sporadic parsing cannot be fixed then I won't bother.

**Comment by andersastor** [comment]
So I'm not sure what changed but when testing it again I'm getting a whole lot of successful parses, way more than 1/s which is more than enough for my use case. Still having the issue with some of the field being private, but I can parse the JSON on my end to bypass this, and do a PR to this repo to fix it properly some time soon (going on vacation soon so won't be until July probably). Closing this issue as all my questions are answered currently.

**Comment by tomasfarias** [comment]
Hey @andersastor. I was just on vacation myself during June, sorry I'm only now reaching back to your questions.

The library should still be considered beta, so some errors are expected. I'm thinking of allowing users of the library to pass their own structures to dictate how the GameState JSON will be parsed. In the meantime, will try to dig into any parsing errors, but any PRs you can open are welcome.

Thank you!

**Comment by tomasfarias** [comment]
Hey @andersastor, quick update: I figured out the issue with JSON parsing. The buffer we were using to read data from the socket was not big enough to accommodate some payloads (sometimes Dota sends 70kb of JSON!). This would cause us to read partial JSONs that would fail parsing. Eventually, we would read a payload small enough so parsing would work again.

As a fix, I just pushed a commit that increases the size of the buffer as we read large JSONs (and also bumped the initial capacity to be at 70kb). I'll publish a new version of the crate that should have your issue solved.

Thanks again for reporting this!
